src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -376,429 +337,218 @@ public int compare(final byte[] o1, final byte[] o2) {
         }
 
         { // next fragment alignment start
-            header.encodingMap.put(EncodingKey.NP_NextFragmentAlignmentStart, ExternalIntegerEncoding.toParam(mateInfoID));
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createGZIP());
+            header.encodingMap.put(EncodingKey.NP_NextFragmentAlignmentStart,
+                    ExternalIntegerEncoding.toParam(exCounter++));
         }
 
         { // template size
-            header.encodingMap.put(EncodingKey.TS_InsetSize, ExternalIntegerEncoding.toParam(mateInfoID));
+            header.externalIds.add(exCounter);
+            header.externalCompressors.put(exCounter, ExternalCompressor.createRANS(RANS.ORDER.ONE));
+            header.encodingMap.put(EncodingKey.TS_InsetSize, ExternalIntegerEncoding.toParam(exCounter));
         }
 
         return header;
     }
 
-    private static int getValue(final EncodingKey key, final CramCompressionRecord record) {
-        switch (key) {
-            case AP_AlignmentPositionOffset:
-                return record.alignmentDelta;
-            case BF_BitFlags:
-                return record.flags;
-            case CF_CompressionBitFlags:
-                return record.compressionFlags;
-            case FN_NumberOfReadFeatures:
-                return record.readFeatures == null ? 0 : record.readFeatures.size();
-            case MF_MateBitFlags:
-                return record.mateFlags;
-            case MQ_MappingQualityScore:
-                return record.mappingQuality;
-            case NF_RecordsToNextFragment:
-                return record.recordsToNextFragment;
-            case NP_NextFragmentAlignmentStart:
-                return record.mateAlignmentStart;
-            case NS_NextFragmentReferenceSequenceID:
-                return record.mateSequenceID;
-            case RG_ReadGroup:
-                return record.readGroupID;
-            case RI_RefId:
-                return record.sequenceId;
-            case RL_ReadLength:
-                return record.readLength;
-            case TC_TagCount:
-                return record.tags == null ? 0 : record.tags.length;
-
-            default:
-                throw new RuntimeException("Unexpected encoding key: " + key.name());
-        }
-    }
-
-    private static void getOptimalIntegerEncoding(final CompressionHeader header, final EncodingKey key, final int minValue,
-                                                  final List<CramCompressionRecord> records) {
-        final IntegerEncodingCalculator calc = new IntegerEncodingCalculator(key.name(), minValue);
+    EncodingDetails buildEncodingForTag(final List<CramCompressionRecord> records, final int id, final int externalID) {
+        @SuppressWarnings("resource") final
+        ByteArrayOutputStream os = new ByteArrayOutputStream(1024 * 1024);
+        final Map<Integer, MutableInt> byteSizes = new HashMap<Integer, MutableInt>();
```

Comment:
`MutableInt` seems... strange... could you explain why you're using it here instead of just Integer?